File: Rules/Extract_Signal_Annotation.py
#----------------------------------------------------------------------------- #
rule extract_signal_annotation:
        input:
            annotation = rules.prepare_annotation.output.outfile,
            wig        = rules.bam2bigWig.output.outfile
        output:
            outfile    = os.path.join(PATH_RDS_TEMP,'{name}','{name}.Extract_Signal_Annotation.rds')
        params:
            peakname    = '{name}',
            scriptdir   = SCRIPT_PATH,
            Rscript     = SOFTWARE['Rscript']['executable']
        log:
            log = os.path.join(PATH_LOG, 'prepare_annotation.log')
        message:"""
                Running: extract_signal_annotation:
                    annot:  {input.annotation}
                    wig:    {input.wig}
                    output: {output.outfile}
            """
        run:
            RunRscript(input, output, params, 'Extract_Signal_Annotation.R')





#----------------------------------------------------------------------------- #
rule extract_signal_peaks:
        input:
            bed        = rules.sort_peak.output.outfile,
            wig        = rules.bam2bigWig.output.outfile
        output:
            outfile    = os.path.join(PATH_RDS_TEMP,'{name}','{name}.Extract_Signal_Peaks.rds')
        params:
            expand_peak = PARAMS['extract_signal']['expand_peak'],
            bin_num     = PARAMS['extract_signal']['bin_num'],
            peakname    = '{name}',
            scriptdir   = SCRIPT_PATH,
            Rscript     = SOFTWARE['Rscript']['executable']
        log:
            log = os.path.join(PATH_LOG, 'prepare_annotation.log')
        message:"""
                Running: extract_signal_peaks:
                    bed: {input.bed}
                    wig: {input.wig}
                    output: {output.outfile}
            """
        run:
            RunRscript(input, output, params, 'Extract_Signal_Peaks.R')


# #----------------------------------------------------------------------------- #
# Annotate Peaks
rule annotate_peaks:
        input:
            annotation = rules.prepare_annotation.output.outfile,
            peaks      = rules.sort_peak.output.outfile
        output:
            outfile    = os.path.join(PATH_RDS_TEMP,'{name}','{name}.Annotate_Peaks.rds')
        params:
            peakname    = '{name}',
            scriptdir   = SCRIPT_PATH,
            Rscript     = SOFTWARE['Rscript']['executable']
        log:
            log = os.path.join(PATH_LOG, 'annotate_peaks.log')
        message:"""
                Running: annotate_peaks:
                    annot: {input.annotation}
                    peaks: {input.peaks}
                    output: {output.outfile}
            """
        run:
            RunRscript(input, output, params, 'Annotate_Peaks.R')

File: pipelines/chipseq/Scripts/Extract_Signal_Annotation.R
# ---------------------------------------------------------------------------- #
#' Extract_Signal_Annotation - given a bigWig file and a set of regions, extracts
#' the signal for downstream analysis
#'
#' @param annotation  - list - processed annotation as given by Prepare_Annotation.R
#' @param wig         - bigWig signal profile
#' @param outfile     - rds output file
#' @param scriptdir   -  location of the R scripts and functions
#'
#' @return saves RDS object with a list of ScoreMatrix and summarized profiles
Extract_Signal_Annotation = function(
    annotation  = NULL, 
    wig         = NULL,
    outfile     = './PEAKS.rds',
    scriptdir   = NULL
){
    
    if(is.null(scriptdir))
        stop('Please specify the script directory')
    
    suppressPackageStartupMessages(library('genomation'))
    suppressPackageStartupMessages(library('GenomicRanges'))
    suppressPackageStartupMessages(library('data.table'))
    source(file.path(scriptdir,'Functions_Helper.R'), local=TRUE)
    # --------------------------------------------------------------- #
    if(is.null(annotation))
        stop('Annotation is not specified')
    
   
    if(is.null(wig))
        stop('wig file is not specified')
    
    # --------------------------------------------------------------- #
    message('Annotation ...')
        annot = readRDS(annotation)
        annot$genomic_annotation$gene = subset(annot$gtf$gtf, type='gene')
        
    message('Extract Profiles ...')
        nams = names(annot$genomic_annotation)
        lsml = lapply(setNames(nams, nams), function(x)
            ScoreMatrixBin(wig, x[[x]], bin.num = bin.num, strand.aware=TRUE))
    
    message('Sumarize Profiles ...')    
        profiles = lapply(lsml, function(x){
            d = data.table(sample = peakname, 
                           signal = rowMeans(x))
        })
    
    lout = list(
        sml      = lsml,
        profiles = profiles
    )
    saveRDS(lout, outfile)
}


# ---------------------------------------------------------------------------- #
Extract_Signal_Annotation(
    annotation  = snakemake@input[['annotation']],
    wig         = snakemake@input[['wig']],
    outfile     = snakemake@output[['outfile']],
    scriptdir   = snakemake@params[['scriptdir']],
    peakname    = snakemake@params[['peakname']]
)